fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/FDBLuceneQueryTest.java

```diff
@@ -0,0 +1,309 @@
+            RecordCursor<FDBQueriedRecord<Message>> primaryKeys;
+            primaryKeys = recordStore.executeQuery(plan);
+            final List<Long> keys = primaryKeys.map(FDBQueriedRecord::getPrimaryKey).map(t -> t.getLong(0)).asList().get();
+            assertEquals(ImmutableSet.of(2L), ImmutableSet.copyOf(keys));
+            if (shouldDeferFetch) {
+                assertLoadRecord(3, context);
+            } else {
+                assertLoadRecord(4, context);
+            }
+        }
+    }
+
+    @ParameterizedTest
+    @BooleanSource
+    public void delayFetchOnUnionOfLuceneScanWithFilters(boolean shouldDeferFetch) throws Exception {
+        try (FDBRecordContext context = openContext()) {
+            openRecordStore(context);
+            final QueryComponent filter1 = new LuceneQueryComponent("civil blood makes civil hands unclean");
+            final QueryComponent filter2 = new LuceneQueryComponent("was king from 966 to 1016");
+            // Query for full records
+            RecordQuery query = RecordQuery.newBuilder()
+                    .setRecordType(TextIndexTestUtils.SIMPLE_DOC)
+                    .setFilter(Query.or(filter1, filter2))
+                    .build();
+            setDeferFetchAfterUnionAndIntersection(shouldDeferFetch);
+            RecordQueryPlan plan = planner.plan(query);
+            Matcher<RecordQueryPlan> matcher = union(
+                            filter(filter1, indexScan(allOf(indexScan("Complex$text_index"),
+                                    indexScanType(IndexScanType.BY_LUCENE),
+                                    bounds(hasTupleString("[[civil blood makes civil hands unclean],[civil blood makes civil hands unclean]]"))))),
+                            filter(filter2, indexScan(allOf(indexScan("Complex$text_index"),
+                                    indexScanType(IndexScanType.BY_LUCENE),
+                                    bounds(hasTupleString("[[was king from 966 to 1016],[was king from 966 to 1016]]"))))),
+                            equalTo(field("doc_id"))
+                    );
+            assertThat(plan, matcher);
+            List<Long> primaryKeys = recordStore.executeQuery(plan).map(FDBQueriedRecord::getPrimaryKey).map(t -> t.getLong(0)).asList().get();
+            assertEquals(ImmutableSet.of(1L, 2L, 4L), ImmutableSet.copyOf(primaryKeys));
+            if (shouldDeferFetch) {
+                assertLoadRecord(5, context);
+            } else {
+                assertLoadRecord(6, context);
+            }
+        }
+    }
+
+    @ParameterizedTest
+    @BooleanSource
+    public void delayFetchOnIntersectionOfLuceneScanWithFilters(boolean shouldDeferFetch) throws Exception {
+        try (FDBRecordContext context = openContext()) {
+            openRecordStore(context);
+            final QueryComponent filter1 = new LuceneQueryComponent("the continuance");
+            final QueryComponent filter2 = new LuceneQueryComponent("grudge");
```

Comment:
There is only one sample that has either of these and it has both, right?

Reply:
Yes, I can add one that has one but not the other. Thats a good point thanks!